plugins/block-dynamic-connection/src/dynamic_list_create.js

```diff
@@ -0,0 +1,124 @@
+Blockly.Blocks['dynamic_list_create'] = {
+  inputCounter: 2,
+  /**
+   * Block for concatenating any number of strings.
+   * @this {Blockly.Block}
+   */
+  init: function() {
+    this.setHelpUrl(Blockly.Msg['LISTS_CREATE_WITH_HELPURL']);
+    this.setStyle('list_blocks');
+    this.appendValueInput('ADD0')
+        .appendField(Blockly.Msg['LISTS_CREATE_WITH_INPUT_WITH']);
+    this.appendValueInput('ADD1');
+    this.setOutput(true, 'Array');
+    this.setTooltip(Blockly.Msg['LISTS_CREATE_WITH_TOOLTIP']);
+  },
+  /**
+   * Create XML to represent number of text inputs.
+   * @return {!Element} XML storage element.
+   * @this {Blockly.Block}
+   */
+  mutationToDom: function() {
+    const container = Blockly.utils.xml.createElement('mutation');
+    const inputNames = this.inputList.map((input) => input.name).join(',');
+    container.setAttribute('inputs', inputNames);
+    container.setAttribute('next', this.inputCounter);
+    return container;
+  },
+  /**
+   * Parse XML to restore the text inputs.
+   * @param {!Element} xmlElement XML storage element.
+   * @this {Blockly.Block}
+   */
+  domToMutation: function(xmlElement) {
+    const items = xmlElement.getAttribute('inputs');
+    if (items) {
+      const inputNames = items.split(',');
+      this.inputList = [];
+      inputNames.forEach((name) => this.appendValueInput(name));
+      this.inputList[0]
+          .appendField(Blockly.Msg['LISTS_CREATE_WITH_INPUT_WITH']);
+    }
+    const next = parseInt(xmlElement.getAttribute('next'));
+    this.inputCounter = next;
+  },
+
+  /**
+   * Check whether a new input should be added and determine where it should go.
+   * @param {Blockly.Connection} connection - The connection that has a
+   * pending connection.
+   * @return {Number} The index before which to insert a new input,
+   * or null if no input should be added.
+   */
+  getIndexForNewInput: function(connection) {
+    if (!connection.targetConnection) {
+      // this connection is available
+      return null;
+    }
+
+    let connectionIndex;
+    for (let i = 0; i < this.inputList.length; i++) {
+      if (this.inputList[i].connection == connection) {
+        connectionIndex = i;
+      }
+    }
+
+    if (connectionIndex == this.inputList.length - 1) {
+      // this connection is the last one and already has a block in it, so
+      // we should add a new connection at the end.
+      return this.inputList.length + 1;
+    }
+
+    const nextInput = this.inputList[connectionIndex + 1];
+    const nextConnection = nextInput && nextInput.connection.targetConnection;
+    if (nextConnection && !nextConnection.sourceBlock_.isInsertionMarker()) {
+      return connectionIndex + 1;
+    }
+
+    // Don't add new connection
+    return null;
+  },
+
+  /**
+   * Called when a block is dragged over one of the connections on this block.
+   * @param {Blockly.Connection} connection - The connection on this block that
```

Comment:
Here and throughout the file, if a type can't be null then there should be a ! in front of it. ex `{!Blockly.Connection}`.